cdap-watchdog/src/main/java/co/cask/cdap/logging/save/LogSaver.java

```diff
@@ -135,17 +141,19 @@ private void createProcessors(Set<Integer> partitions) throws Exception {
     }
   }
 
-  private void cancelLogCollectorCallbacks() {
+  private Set<Cancellable> cancelLogCollectorCallbacks() {
+    Set<Cancellable> cancellables = new HashSet<>();
    for (Entry<Integer, Cancellable> entry : kafkaCancelMap.entrySet()) {
       if (entry.getValue() != null) {
         LOG.info("Cancelling kafka callback for partition {}", entry.getKey());
         kafkaCancelCallbackLatchMap.get(entry.getKey()).countDown();
-        entry.getValue().cancel();
+        cancellables.add(entry.getValue());
       }
     }
 
     kafkaCancelMap.clear();
     kafkaCancelCallbackLatchMap.clear();
```

Comment:
This class need refactoring to separate out the logic that handles Kafka partition consumer and its callbacks. I have added a TODO with JIRA CDAP-6475 for the refactoring. For now I'll leave it as-is. What do you say?